lib/driver-cli.js

```diff
@@ -0,0 +1,247 @@
+import _ from 'lodash';
+import { KNOWN_DRIVERS, validateDriversYaml, writeInstalledDriver,
+         INSTALL_TYPE_LOCAL, INSTALL_TYPE_NPM, INSTALL_TYPE_GIT,
+         INSTALL_TYPE_GITHUB } from './drivers';
+import { getLatestVersion, installPackage, linkPackage } from './npm';
+import { errAndQuit, log, spinWith } from './cli-utils';
+import { util } from 'appium-support';
+
+let INSTALLED_DRIVERS;
+
+async function runDriverCommand (args) {
+  // TODO driver.json file should be locked while any of these commands are
+  // running to prevent weird situations
+  let jsonResult = null;
+  const {json, appiumHome} = args;
+  const logFn = (msg) => log(json, msg);
+  try {
+    INSTALLED_DRIVERS = await validateDriversYaml(appiumHome, logFn);
+    switch (args.driverCommand) {
+      case 'list':
+        jsonResult = await list(args);
+        break;
+      case 'install':
+        jsonResult = await install(args);
+        break;
+      case 'uninstall':
+        jsonResult = await uninstall(args);
+        break;
+      case 'update':
+        jsonResult = await update(args);
+        break;
+      case 'verify':
+        // TODO create a 'verify' command that just checks driver.json and then
+        // that everything in driver.json can be required OK.
+        // verify could also have a --reinstall flag which installs any drivers
+        // which are listed in the drivers.json but which cannot be required
+        // for whatever reason
+        break;
+      default:
+        throw new Error(`Cannot handle driver command ${args.driverCommand}`);
+    }
+  } catch (err) {
+    errAndQuit(json, err);
+  }
+
+  if (json) {
+    console.log(JSON.stringify(jsonResult));
+  }
+}
+
+async function list ({appiumHome, json, showInstalled, showUpdates}) {
+  const lsMsg = `Listing ${showInstalled ? 'installed' : 'available'} drivers`;
+  const drivers = {};
+
+  for (const name of Object.keys(INSTALLED_DRIVERS)) {
+    drivers[name] = INSTALLED_DRIVERS[name];
+    drivers[name].installed = true;
+  }
+
+  if (!showInstalled) {
+    for (const name of Object.keys(KNOWN_DRIVERS)) {
+      if (!drivers[name]) {
+        drivers[name] = {
+          pkgName: KNOWN_DRIVERS[name],
+          installed: false,
+        };
+      }
+    }
+  }
+
+  await spinWith(json, lsMsg, async () => {
+    if (showUpdates) {
+      for (const [, data] of _.toPairs(drivers)) {
```

Comment:
you can use _.values(drivers) if keys are not needed

Reply:
`for (const {version, pkgName, installType} of _.values(drivers))`

Reply:
I’m actually mutating the value of drivers though, so I need access to it.